File: ext-go-legacy/internal/features/socketclient/payloads/payloads.go
// Package payloads holds the Go counterparts of the PHP socket-client payload
// objects (SConcur\Features\SocketClient\Payloads\*) and the connection metadata the
// client streams back to PHP. Struct tags are the short keys exchanged via
// MessagePack.
//
// Every message is a command envelope (cm/p) under MethodSocketClient — mirrors the
// HTTP-client feature: cm selects the sub-operation, p carries that command's
// parameters (decoded into ConnectParams / SendParams / CloseParams).
package payloads

import (
	"sconcur/internal/types"

	"github.com/vmihailenco/msgpack/v5"
)

// Envelope is the command envelope decoded from the msgpack message.
// PHP: SConcur\Features\SocketClient\Payloads\Base\BaseSocketClientPayload (cm from
// the command enum, p from the parameters' getData()).
type Envelope struct {
	Command types.SocketClientCommand `json:"cm" msgpack:"cm"`
	Params  msgpack.RawMessage        `json:"p" msgpack:"p"`
}

// ConnectParams is the `p` content of a Connect command — the remote address and the
// per-connection tuning (timeouts in milliseconds, sizes in bytes). Defaults are
// supplied by the PHP side (SocketClientOptions). There is no single "operation
// time": ConnectTimeoutMs bounds the dial, ReadTimeoutMs the idle wait for an
// inbound frame, WriteTimeoutMs one frame write — the mandatory execution bounds for
// a long-lived connection.
// PHP: SConcur\Features\SocketClient\Payloads\ConnectPayloadParameters.
type ConnectParams struct {
	Address string `json:"ad" msgpack:"ad"`
	// ConnectTimeoutMs bounds establishing the TCP connection (net.Dialer). 0 falls
	// back to the default.
	ConnectTimeoutMs int `json:"ct" msgpack:"ct"`
	// ReadTimeoutMs is the idle timeout waiting for the next inbound frame in read().
	// 0 disables it (a connection may stay idle forever).
	ReadTimeoutMs int `json:"rt" msgpack:"rt"`
	// WriteTimeoutMs bounds writing one outbound frame.
	WriteTimeoutMs int `json:"wt" msgpack:"wt"`
	// MaxMessageBytes caps the length of a single inbound frame (guards against a huge
	// length prefix).
	MaxMessageBytes int `json:"mmb" msgpack:"mmb"`
}

// SendParams is the `p` content of a Send command: the connection to write to and
// the frame bytes (binary-safe).
// PHP: SConcur\Features\SocketClient\Payloads\SendPayload.
type SendParams struct {
	ConnectionId string `json:"cid" msgpack:"cid"`
	Data         string `json:"dt" msgpack:"dt"`
}

// CloseParams is the `p` content of a Close command: the connection to close.
// PHP: SConcur\Features\SocketClient\Payloads\ClosePayload.
type CloseParams struct {
	ConnectionId string `json:"cid" msgpack:"cid"`
}

// ConnectionMeta is the first result the client emits for a Connect: the connection
// id (used to route Send/Close) plus the resolved addresses. Subsequent results
// (pulled via next) are raw inbound frames, not this struct.
// PHP: decoded in SConcur\Features\SocketClient\SocketClient::connect.
type ConnectionMeta struct {
	ConnectionId string `json:"cid" msgpack:"cid"`
	RemoteAddr   string `json:"ra" msgpack:"ra"`
	LocalAddr    string `json:"la" msgpack:"la"`
}

File: ext/src/features/socketclient/payloads.rs
//! Mirrors ext-go-legacy/internal/features/socketclient/payloads.
//!
//! Every message is a command envelope (`cm`/`p`) under the socketClient method:
//! `cm` selects the sub-operation, `p` carries that command's parameters. `p` is
//! a nested map, not a blob of its own — `'p' => $this->getParameters()->getData()`
//! on the PHP side — so it is decoded as a value and typed per command.

use serde::Deserialize;

#[derive(Deserialize)]
pub struct Envelope {
    #[serde(rename = "cm", default)]
    pub command: String,
    #[serde(rename = "p", default = "nil_value")]
    pub params: rmpv::Value,
}

fn nil_value() -> rmpv::Value {
    rmpv::Value::Nil
}

/// The `p` of a Connect — the remote address and the per-connection tuning.
/// There is no single "operation time": the connect timeout bounds the dial, the
/// read timeout the idle wait for an inbound frame, the write timeout one frame.
/// PHP: SConcur\Features\SocketClient\Payloads\ConnectPayloadParameters.
#[derive(Deserialize, Default)]
pub struct ConnectParams {
    #[serde(rename = "ad", default)]
    pub address: String,
    #[serde(rename = "ct", default)]
    pub connect_timeout_ms: i64,
    /// 0 disables it: a connection may stay idle forever.
    #[serde(rename = "rt", default)]
    pub read_timeout_ms: i64,
    #[serde(rename = "wt", default)]
    pub write_timeout_ms: i64,
    #[serde(rename = "mmb", default)]
    pub max_message_bytes: i64,
}

/// The `p` of a Send: the connection to write to and the frame bytes.
#[derive(Deserialize)]
pub struct SendParams {
    #[serde(rename = "cid", default)]
    pub connection_id: String,
    /// Binary-safe, so an untyped value rather than a String.
    #[serde(rename = "dt", default = "nil_value")]
    pub data: rmpv::Value,
}

impl SendParams {
    /// The frame's bytes, taken out of the params rather than copied out of
    /// them: the decoded value already owns them and nothing reads `data`
    /// again.
    pub fn take_data_bytes(&mut self) -> Vec<u8> {
        match std::mem::replace(&mut self.data, rmpv::Value::Nil) {
            rmpv::Value::String(text) => text.into_bytes(),
            rmpv::Value::Binary(bytes) => bytes,
            _ => Vec::new(),
        }
    }
}

/// The `p` of a Close: the connection to close.
#[derive(Deserialize, Default)]
pub struct CloseParams {
    #[serde(rename = "cid", default)]
    pub connection_id: String,
}

/// The first result a Connect emits: the id used to route Send/Close, plus the
/// resolved addresses. Every result after it is a raw inbound frame.
pub struct ConnectionMeta {
    pub connection_id: String,
    pub remote_addr: String,
    pub local_addr: String,
}

impl ConnectionMeta {
    pub fn encode(&self) -> Vec<u8> {
        use rmp::encode;

        let mut buffer = Vec::with_capacity(96);

        let _ = encode::write_map_len(&mut buffer, 3);

        for (key, value) in [
            ("cid", &self.connection_id),
            ("ra", &self.remote_addr),
            ("la", &self.local_addr),
        ] {
            let _ = encode::write_str(&mut buffer, key);
            let _ = encode::write_str(&mut buffer, value);
        }

        buffer
    }
}
